File: app/src/main/java/com/sumian/hw/improve/report/ReportFragment.java
package com.sumian.hw.improve.report;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.support.v4.app.FragmentManager;
import android.support.v4.content.LocalBroadcastManager;
import android.view.View;

import com.blankj.utilcode.util.LogUtils;
import com.sumian.hw.base.HwBaseFragment;
import com.sumian.hw.improve.guideline.dialog.ReportGuidelineDialog;
import com.sumian.hw.improve.guideline.utils.GuidelineUtils;
import com.sumian.hw.improve.main.bean.PushReport;
import com.sumian.hw.improve.report.calendar.CalendarDialog;
import com.sumian.hw.improve.report.dailyreport.DailyReportFragment;
import com.sumian.hw.improve.report.weeklyreport.WeeklyReportFragment;
import com.sumian.hw.improve.widget.TabIndicatorView;
import com.sumian.hw.log.LogManager;
import com.sumian.hw.push.ReportPushManager;
import com.sumian.hw.utils.FragmentUtil;
import com.sumian.sd.R;
import com.sumian.sd.app.AppManager;
import com.sumian.sd.main.OnEnterListener;

/**
 * Created by sm
 * on 2018/3/6.
 * desc:
 */

public class ReportFragment extends HwBaseFragment implements TabIndicatorView.OnSwitchIndicatorCallback, OnEnterListener {

    TabIndicatorView mTabIndicatorView;

    private String[] mFTags = {DailyReportFragment.class.getSimpleName(), WeeklyReportFragment.class.getSimpleName()};
    private int mCurrentPosition = 0;

    private BroadcastReceiver mBroadcastReceiver;
    private ReportGuidelineDialog mReportGuidelineDialog;

    public static ReportFragment newInstance() {
        return new ReportFragment();
    }

    @SuppressWarnings("ConstantConditions")
    @Override
    public void onEnter() {
        LogManager.appendUserOperationLog("点击进入 '报告' 页面");
        // check job scheduler
        AppManager.getJobScheduler().checkJobScheduler();
        // check guideline
        if (GuidelineUtils.needShowDailyUserGuide()) {
            mReportGuidelineDialog = new ReportGuidelineDialog(getContext());
            mReportGuidelineDialog.show();
        }
        // check push report
        boolean showPushReport = showPushReportInNeeded();
        LogUtils.d(mTabIndicatorView);
        if (!showPushReport && mCurrentPosition != -1 && mTabIndicatorView != null) {
            mTabIndicatorView.selectTabByPosition(mCurrentPosition);
        }
    }

    /**
     * 如果有push report，则切换到对应的tab
     *
     * @return 是否有push report存在
     */
    private boolean showPushReportInNeeded() {
        PushReport pushReport = ReportPushManager.getInstance().getPushReport();
        if (pushReport != null) {
            if (pushReport.getPushType() == PushReport.PUSH_TYPE_DAILY_REPORT) {
                mTabIndicatorView.selectTabByPosition(0);
            } else {
                mTabIndicatorView.selectTabByPosition(1);
            }
            return true;
        }
        return false;
    }

    @Override
    protected int getLayoutId() {
        return R.layout.hw_fragment_main_report;
    }

    @SuppressWarnings("ConstantConditions")
    @Override
    protected void initWidget(View root) {
        super.initWidget(root);
        LogUtils.d();
        mTabIndicatorView = root.findViewById(R.id.tab_indicator_view);
        mTabIndicatorView.setOnSwitchIndicatorCallback(this);
    }

    @SuppressWarnings("ConstantConditions")
    @Override
    protected void initData() {
        super.initData();
        IntentFilter filter = new IntentFilter();
        filter.addAction(CalendarDialog.ACTION_SELECT_DATE);
        LocalBroadcastManager.getInstance(getContext()).registerReceiver(mBroadcastReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                switch (intent.getAction()) {
                    case CalendarDialog.ACTION_SELECT_DATE:
                        mTabIndicatorView.updateTabsUiBySelectPosition(0);
                        onSwitchIndicator(null, 0);
                        break;
                    default:
                        break;
                }
            }
        }, filter);
    }

    @Override
    public void onHiddenChanged(boolean hidden) {
        super.onHiddenChanged(hidden);
        if (hidden) {
            if (mReportGuidelineDialog != null && mReportGuidelineDialog.isShowing()) {
                mReportGuidelineDialog.dismiss();
            }
        }
    }

    @SuppressWarnings("ConstantConditions")
    @Override
    protected void onRelease() {
        super.onRelease();
        LocalBroadcastManager.getInstance(getContext()).unregisterReceiver(mBroadcastReceiver);
    }

    @SuppressWarnings("ConstantConditions")
    @Override
    public void onSwitchIndicator(View v, int position) {
        LogManager.appendPhoneLog("com.sumian.app.improve.report.ReportFragment.onSwitchIndicator position: " + position);
        showFragmentByPosition(position);
        mCurrentPosition = position;
    }

    private void showFragmentByPosition(int position) {
        FragmentManager fragmentManager = getFragmentManager();
        if (fragmentManager == null) {
            return;
        }
        FragmentUtil.Companion.switchFragment(R.id.report_fragment_container, fragmentManager, mFTags, position,
                p -> {
                    if (p == 0) {
                        return DailyReportFragment.newInstance();
                    } else {
                        return WeeklyReportFragment.newInstance();
                    }
                }, new FragmentUtil.DefaultRunOnCommitCallbackImpl()
        );
    }

    @SuppressWarnings("ConstantConditions")
    @Override
    public void onShowCalendar(View v) {
        CalendarDialog.create(getCurrentShowReportTime()).show(getFragmentManager(), CalendarDialog.class.getSimpleName());
    }

    private long getCurrentShowReportTime() {
        FragmentManager fragmentManager = getFragmentManager();
        if (fragmentManager == null) {
            return 0;
        }
        DailyReportFragment dailyReportFragment = (DailyReportFragment) fragmentManager.findFragmentByTag(mFTags[0]);
        if (dailyReportFragment == null) {
            return 0;
        }
        return dailyReportFragment.getCurrentReportTime();
    }
}

File: app/src/main/java/com/sumian/hw/report/ReportFragment.kt
package com.sumian.hw.report

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.support.v4.app.Fragment
import android.support.v4.content.LocalBroadcastManager
import android.view.View
import com.sumian.hw.base.HwBasePresenter
import com.sumian.hw.guideline.dialog.ReportGuidelineDialog
import com.sumian.hw.guideline.utils.GuidelineUtils
import com.sumian.hw.log.LogManager
import com.sumian.hw.push.ReportPushManager
import com.sumian.hw.report.bean.PushReport
import com.sumian.hw.report.calendar.CalendarDialog
import com.sumian.hw.report.fragment.DailyReportFragment
import com.sumian.hw.report.fragment.WeeklyReportFragment
import com.sumian.hw.report.widget.tab.TabIndicatorView
import com.sumian.hw.utils.FragmentUtil
import com.sumian.sd.R
import com.sumian.sd.app.AppManager
import com.sumian.sd.main.OnEnterListener
import com.sumian.sd.theme.three.base.SkinBaseFragment
import com.sumian.sd.theme.three.loader.SkinManager
import kotlinx.android.synthetic.main.hw_fragment_main_report.*
import kotlinx.android.synthetic.main.hw_lay_tab_indicator_view.view.*

@Suppress("UNREACHABLE_CODE")
/**
 * Created by sm
 * on 2018/3/6.
 * desc:
 */

class ReportFragment : SkinBaseFragment<HwBasePresenter>(), TabIndicatorView.OnSwitchIndicatorCallback, OnEnterListener {

    companion object {

        fun newInstance(): ReportFragment {
            return ReportFragment()
        }

    }

    private var mCurrentPosition = 0

    private val mFTags by lazy {
        arrayOf(DailyReportFragment::class.java.simpleName, WeeklyReportFragment::class.java.simpleName)
    }

    private val mBroadcastReceiver: BroadcastReceiver  by lazy {
        object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                when (intent.action) {
                    CalendarDialog.ACTION_SELECT_DATE -> {
                        tab_indicator_view.updateTabsUiBySelectPosition(0)
                        onSwitchIndicator(tab_indicator_view, 0)
                    }
                }
            }
        }
    }

    private val mReportGuidelineDialog: ReportGuidelineDialog  by lazy {
        ReportGuidelineDialog(context!!)
    }

    private val currentShowReportTime: Long
        get() {
            val fragmentManager = fragmentManager ?: return 0
            val dailyReportFragment = fragmentManager.findFragmentByTag(mFTags[0]) as DailyReportFragment
            return dailyReportFragment.currentReportTime
        }

    override fun getLayoutId(): Int {
        return R.layout.hw_fragment_main_report
    }

    override fun initWidget(root: View) {
        super.initWidget(root)
        tab_indicator_view?.let {
            it.setOnSwitchIndicatorCallback(this)
            dynamicAddView(it, "background", R.color.n1_color_day)
        }
    }

    override fun initData() {
        super.initData()
        val filter = IntentFilter()
        filter.addAction(CalendarDialog.ACTION_SELECT_DATE)
        LocalBroadcastManager.getInstance(activity!!).registerReceiver(mBroadcastReceiver, filter)
        onEnter(null)
    }

    override fun onHiddenChanged(hidden: Boolean) {
        super.onHiddenChanged(hidden)
        if (hidden) {
            if (mReportGuidelineDialog.isShowing) {
                mReportGuidelineDialog.dismiss()
            }
        }
    }

    override fun onRelease() {
        super.onRelease()
        LocalBroadcastManager.getInstance(activity!!).unregisterReceiver(mBroadcastReceiver)
    }

    override fun onEnter(data: String?) {
        LogManager.appendUserOperationLog("点击进入 '报告' 页面")
        // check job scheduler
        AppManager.getJobScheduler().checkJobScheduler()
        // check guideline
        if (GuidelineUtils.needShowDailyUserGuide()) {
            mReportGuidelineDialog.show()
        }
        tab_indicator_view?.iv_back?.visibility = if (SkinManager.getInstance().isNightMode) View.INVISIBLE else View.VISIBLE
    }

    override fun onBack(v: View) {
        activity?.finish()
    }

    override fun onSwitchIndicator(v: View, position: Int) {
        LogManager.appendPhoneLog("com.sumian.app.improve.report.ReportFragment.onSwitchIndicator position: $position")
        showFragmentByPosition(position)
        mCurrentPosition = position
    }

    override fun onShowCalendar(v: View) {
        CalendarDialog.create(currentShowReportTime).show(fragmentManager!!, CalendarDialog::class.java.simpleName)
    }

    private fun showFragmentByPosition(position: Int) {
        FragmentUtil.switchFragment(R.id.report_fragment_container,
                fragmentManager!!,
                mFTags,
                position,
                object : FragmentUtil.FragmentCreator {
                    override fun createFragmentByPosition(position: Int): Fragment {
                        return if (position == 0) {
                            DailyReportFragment.newInstance()
                        } else {
                            WeeklyReportFragment.newInstance()
                        }
                    }
                }
                , FragmentUtil.DefaultRunOnCommitCallbackImpl())
    }
}
